node/libs/crypto/src/bn254/mod.rs

```diff
@@ -0,0 +1,200 @@
+    pub fn public(&self) -> PublicKey {
+        let pk = bn254::PublicKey::from_private_key(&self.0);
+        PublicKey(pk)
+    }
+}
+
+impl ByteFmt for SecretKey {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        bn254::PrivateKey::try_from(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("Failed to decode secret key: {e:?}"))
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        self.0.to_bytes().unwrap()
+    }
+}
+
+/// Type safety wrapper around a `bn254` public key.
+#[derive(Clone)]
+pub struct PublicKey(bn254::PublicKey);
+
+impl PartialEq for PublicKey {
+    fn eq(&self, other: &Self) -> bool {
+        ByteFmt::encode(self).eq(&ByteFmt::encode(other))
+    }
+}
+
+impl Eq for PublicKey {}
+
+impl ByteFmt for PublicKey {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        bn254::PublicKey::from_compressed(bytes)
+            .map(Self)
+            .map_err(|err| anyhow!("Error decoding public key: {err:?}"))
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        self.0.to_compressed().unwrap()
+    }
+}
+
+impl std::hash::Hash for PublicKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write(&self.0.to_compressed().unwrap())
+    }
+}
+
+impl PartialOrd for PublicKey {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for PublicKey {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        ByteFmt::encode(self).cmp(&ByteFmt::encode(other))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Signature(bn254::Signature);
+
+impl Signature {
+    pub fn verify(&self, msg: &[u8], pk: &PublicKey) -> Result<(), Error> {
+        let result = ECDSA::verify(msg, &self.0, &pk.0);
+        match result {
+            Ok(()) => Ok(()),
+            Err(err) => Err(Error::SignatureVerification(err)),
+        }
+    }
+}
+
+impl PartialEq for Signature {
+    fn eq(&self, other: &Self) -> bool {
+        ByteFmt::encode(self).eq(&ByteFmt::encode(other))
+    }
+}
+
+impl Eq for Signature {}
+
+impl ByteFmt for Signature {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        bn254::Signature::from_compressed(bytes)
+            .map(Self)
+            .map_err(|err| anyhow!("Error decoding signature: {err:?}"))
+    }
+    fn encode(&self) -> Vec<u8> {
+        self.0.to_compressed().unwrap()
+    }
+}
+
+impl PartialOrd for Signature {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Signature {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        ByteFmt::encode(self).cmp(&ByteFmt::encode(other))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct AggregateSignature(Signature);
+
+impl AggregateSignature {
+    pub fn aggregate<'a>(sigs: impl IntoIterator<Item = &'a Signature>) -> Self {
+        let sigs: Vec<bn254::Signature> = sigs.into_iter().map(|s| s.0).collect();
```

Comment:
Instead of building a vector, you can take the first element using next() and then iterate over the remaining elements.

Reply:
Fixed.